src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
```

Comment:
Agreed. 
I guess, when you do not have to go through reflection, the stack probing can be fast.